File: 10. Dynamic Programming/02. Longest Common Subsequence/LCS.py
def LCS_recursive(s1, s2, i, j):
    if i < 0 or j < 0:
        return 0
    
    if s1[i] == s2[j]:
        return 1 + LCS_recursive(s1, s2, i - 1, j - 1)
    else:
        return max(LCS_recursive(s1, s2, i - 1, j), LCS_recursive(s1, s2, i, j - 1))


def LCS_memoization(s1, s2, i, j, dp):
    if i < 0 or j < 0:
        return 0
    
    if dp[i][j] != -1:
        return dp[i][j]
    
    if s1[i] == s2[j]:
        dp[i][j] = 1 + LCS_memoization(s1, s2, i - 1, j - 1, dp)
    else:
        dp[i][j] = max(LCS_memoization(s1, s2, i - 1, j, dp), LCS_memoization(s1, s2, i, j - 1, dp))
    
    return dp[i][j]


def LCS_tabulation(s1, s2):
    n, m = len(s1), len(s2)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    
    return dp[n][m]


def LCS_tabulation_space_optimized(s1, s2):
    n, m = len(s1), len(s2)
    prev = [0] * (m + 1)
    cur = [0] * (m + 1)
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                cur[j] = 1 + prev[j - 1]
            else:
                cur[j] = max(prev[j], cur[j - 1])
        prev = cur[:]
    
    return prev[m]


def LCS_print(s1, s2):
    n, m = len(s1), len(s2)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    
    lcs = []
    i, j = n, m
    while i > 0 and j > 0:
        if s1[i - 1] == s2[j - 1]:
            lcs.append(s1[i - 1])
            i -= 1
            j -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1
    
    lcs.reverse()
    return ''.join(lcs)


def main():
    s1 = "abcde"
    s2 = "ace"
    
    # Brute Force Recursive
    print("Length of LCS (Brute Force):", LCS_recursive(s1, s2, len(s1) - 1, len(s2) - 1))
    
    # Memoization
    dp = [[-1 for _ in range(len(s2))] for _ in range(len(s1))]
    print("Length of LCS (Memoization):", LCS_memoization(s1, s2, len(s1) - 1, len(s2) - 1, dp))
    
    # Tabulation
    print("Length of LCS (Tabulation):", LCS_tabulation(s1, s2))
    
    # Space Optimized Tabulation
    print("Length of LCS (Tabulation Space Optimized):", LCS_tabulation_space_optimized(s1, s2))
    
    # LCS Printing
    print("The LCS is:", LCS_print(s1, s2))

main()

File: 10. Dynamic Programming/02. Longest Common Subsequence/LCS.go
package main

import (
	"fmt"
)

// 1. Brute Force Recursive
func LCS(s1, s2 string, i, j int) int {
	if i < 0 || j < 0 {
		return 0
	}
	if s1[i] == s2[j] {
		return 1 + LCS(s1, s2, i-1, j-1)
	}
	return max(LCS(s1, s2, i-1, j), LCS(s1, s2, i, j-1))
}

// 2. Memoization
func LCSMemo(s1, s2 string, i, j int, dp [][]int) int {
	if i < 0 || j < 0 {
		return 0
	}
	if dp[i][j] != -1 {
		return dp[i][j]
	}
	if s1[i] == s2[j] {
		dp[i][j] = 1 + LCSMemo(s1, s2, i-1, j-1, dp)
	} else {
		dp[i][j] = max(LCSMemo(s1, s2, i-1, j, dp), LCSMemo(s1, s2, i, j-1, dp))
	}
	return dp[i][j]
}

// 3. Tabulation
func LCSTab(s1, s2 string) int {
	n, m := len(s1), len(s2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}
	return dp[n][m]
}

// 4. Space Optimized
func LCSSpaceOptimized(s1, s2 string) int {
	n, m := len(s1), len(s2)
	prev := make([]int, m+1)
	cur := make([]int, m+1)

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				cur[j] = 1 + prev[j-1]
			} else {
				cur[j] = max(prev[j], cur[j-1])
			}
		}
		copy(prev, cur)
	}
	return prev[m]
}

// 5. LCS Print
func LCSPrint(s1, s2 string) string {
	n, m := len(s1), len(s2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}

	i, j := n, m
	lcs := []byte{}

	for i > 0 && j > 0 {
		if s1[i-1] == s2[j-1] {
			lcs = append(lcs, s1[i-1])
			i--
			j--
		} else if dp[i-1][j] > dp[i][j-1] {
			i--
		} else {
			j--
		}
	}

	// reverse lcs
	for i := 0; i < len(lcs)/2; i++ {
		lcs[i], lcs[len(lcs)-1-i] = lcs[len(lcs)-1-i], lcs[i]
	}
	return string(lcs)
}

// Utility
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// MAIN
func main() {
	s1 := "abcde"
	s2 := "ace"

	fmt.Println("Brute Force:", LCS(s1, s2, len(s1)-1, len(s2)-1))

	dp := make([][]int, len(s1))
	for i := range dp {
		dp[i] = make([]int, len(s2))
		for j := range dp[i] {
			dp[i][j] = -1
		}
	}
	fmt.Println("Memoization:", LCSMemo(s1, s2, len(s1)-1, len(s2)-1, dp))
	fmt.Println("Tabulation:", LCSTab(s1, s2))
	fmt.Println("Space Optimized:", LCSSpaceOptimized(s1, s2))
	fmt.Println("LCS String:", LCSPrint(s1, s2))
}
